Use ImageProviderInterface instead of ImageProvider for contructor's …

jerome-fix opened the pull request:
…signature

<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject



I need to use my own ImageProvider and FileProvider for differents reasons.
Since these base providers are final classes (or declared as final) I can't extend them anymore.

Here my config : 
```
parameters:
  sonata.media.provider.image.class: App\Media\Provider\ImageProvider
  sonata.media.manager.media.class:  App\Manager\MediaManager
  sonata.media.provider.file.class: App\Media\Provider\FileProvider
```

  
Since the introduction of ImageUploadDimension validator I can't upload any new media, because the signature of the constructor expects an Sonata\MediaBundle\Provider\ImageProvider.

So I changed the constructor's signature to take an ImageProviderInterface instead. 

My provider can implement this interface as Sonata\MediaBundle\ImageProvide

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 3.x is for everything backwards compatible, like patches, features and deprecation notices
    - master is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataMediaBundle/blob/3.x/CONTRIBUTING.md#the-base-branch
-->
 

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
--> 

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataMediaBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Added ImageProviderInterface


### Deprecated
 - Deprecations about the "sonata.media.provider.image" service alias,use "Sonata\MediaBundle\Provider\ImageProviderInterface" instead.

```

<!--
    If this is a work in progress, uncomment the "To do" section.
    You can add as many tasks as you want.
    If some are not relevant, just remove them.
-->
<!--
## To do

- [ ] Update the tests;
- [ ] Update the documentation;
- [ ] Add an upgrade note.
-->

wbloszyk commented:
I see next problem with extends FileProvider which will be final too:
``` php
class ImageProvider extends FileProvider {}
```
IMHO We should have:
- BaseProvider implements MediaProviderInterface
- BaseFileProvider extends BaseProvider implements FileProviderInterface
- BaseImageProvider extends BaseProvider implements ImageProviderInterface
- BaseVideoProvider extends BaseProvider implements VideoProviderInterface

Technicly we can also add `DisplayableProviderInterface` but I don't think it will be useful for now. Mybe we can add to `DimensionValidator`? WDYT? @core23

VincentLanglet commented:
> I see next problem with extends FileProvider which will be final too:
> 
> ```
> class ImageProvider extends FileProvider {}
> ```
> 
> IMHO We should have:
> 
> * BaseProvider implements MediaProviderInterface
> * BaseFileProvider extends BaseProvider implements FileProviderInterface
> * BaseImageProvider extends BaseProvider implements ImageProviderInterface
> * BaseVideoProvider extends BaseProvider implements VideoProviderInterface
> 
> Do you want to do it? @jerome-fix
> 
> Technicly we can also add `DisplayableProviderInterface` but I don't think it will be useful for now. Mybe we can add to `DimensionValidator`? WDYT? @core23

This can be in another PR.

VincentLanglet commented on src/Provider/ImageProviderInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+namespace Sonata\MediaBundle\Provider;
+
+interface ImageProviderInterface extends MediaProviderInterface
```
extends FileProviderInterface ?

wbloszyk replied:
`ImageProvider extends FileProvider` so `ImageProviderInterface extends FileProviderInterface`

wbloszyk commented on src/Provider/ImageProviderInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+namespace Sonata\MediaBundle\Provider;
+
+interface ImageProviderInterface extends MediaProviderInterface
```
```suggestion
interface ImageProviderInterface extends FileProviderInterface
```

jerome-fix replied:
Makes sense. I'll change it.

jerome-fix commented:
> I see next problem with extends FileProvider which will be final too:
> 
> ```
> class ImageProvider extends FileProvider {}
> ```
> 
> IMHO We should have:
> 
>     * BaseProvider implements MediaProviderInterface
> 
>     * BaseFileProvider extends BaseProvider implements FileProviderInterface
> 
>     * BaseImageProvider extends BaseProvider implements ImageProviderInterface
> 
>     * BaseVideoProvider extends BaseProvider implements VideoProviderInterface
> 
> 
> Do you want to do it? @jerome-fix
> 
> Technicly we can also add `DisplayableProviderInterface` but I don't think it will be useful for now. Mybe we can add to `DimensionValidator`? WDYT? @core23

wbloszyk commented on src/Validator/Constraints/ImageUploadDimensionValidator.php:
```diff
@@ -16,6 +16,7 @@
 use Imagine\Image\ImagineInterface;
 use Sonata\MediaBundle\Model\MediaInterface;
 use Sonata\MediaBundle\Provider\ImageProvider;
```
```suggestion
```

wbloszyk commented on src/Validator/Constraints/ImageUploadDimensionValidator.php:
```diff
@@ -32,7 +33,7 @@ final class ImageUploadDimensionValidator extends ConstraintValidator
      */
     private $imageProvider;
```
Change @var to `ImageProviderInterface` too

jordisala1991 commented on src/Provider/ImageProviderInterface.php:
```diff
@@ -0,0 +1,22 @@
+
+namespace Sonata\MediaBundle\Provider;
+
+interface ImageProviderInterface extends FileProviderInterface
```
Shouldnt the FileProvider have the FileProviderInterface instead of here extending it? What is the better option?

wbloszyk replied:
In ImageProvider using of FileProviderInterface is possible. People who write own ImageProvider should be force to implement it becouse Image is File too.

jerome-fix commented:
Hi,

Do I need to make adjustments for this request? or is that okay for now? 

I agree with @wbloszyk with the following proposals: 

- BaseProvider implements MediaProviderInterface
- BaseFileProvider extends BaseProvider implements FileProviderInterface
- BaseImageProvider extends BaseProvider implements ImageProviderInterface
- BaseVideoProvider extends BaseProvider implements VideoProviderInterface
    
I may be able to suggest a PR in a few days if you want.

Jérôme.

VincentLanglet commented:
> Do I need to make adjustments for this request? or is that okay for now?

@jerome-fix I think you just have to update your PR with the @phansys request:

> Could you please update the related tests in order to use this interface?
> By instance
> 
> https://github.com/sonata-project/SonataMediaBundle/blob/8bee9ef451d9041a37ee5c46803190a35a197a01/tests/Metadata/ProxyMetadataBuilderTest.php#L56

phansys commented:
Maybe we could use this interface as alias for service "sonata.media.provider.image", but I think this change should be done in the `master` branch since it should be considered as a BC break.
https://github.com/sonata-project/SonataMediaBundle/blob/e9411ad39543b3de02b5ff8a22af9072da3d77c4/src/Resources/config/provider.xml#L114

phansys commented:
IMO, we should create a service alias "Sonata\MediaBundle\Provider\ImageProviderInterface" pointing to "sonata.media.provider.image", and deprecate the alias "%sonata.media.provider.image.class%" in favor of new one.
@sonata-project/contributors, WDYT?

jerome-fix commented:
Hi @phansys ,

I don't understand your "minor" tag.
This is blocking for all those who have their own ImageProvider.

Is there any possibility of merging this pull request quickly ?  Or you're waiting for @jordisala1991 's validation ?

Jérôme.

VincentLanglet commented:
> I don't understand your "minor" tag.
> This is blocking for all those who have their own ImageProvider.

This will be release as a minor version because it's a new feature. 3.27.

> IMO, we should create a service alias "Sonata\MediaBundle\Provider\ImageProviderInterface" pointing to "sonata.media.provider.image", and deprecate the alias "%sonata.media.provider.image.class%" in favor of new one.
@sonata-project/contributors, WDYT?

@phansys Is it a blocking point or can we merge this ?

phansys commented:
I think it's not a blocker, since AFAIK the autowiring layer is exposed only by courtesy.
I guess the required changes to provide what I've suggested should be covered by the following lines:
```xml
<service id="%sonata.media.provider.image.class%" alias="sonata.media.provider.image">
    <deprecated>The "%alias_id%" service alias is deprecated since sonata-project/media-bundle 3.x and will be removed in version 4.0. Use "Sonata\MediaBundle\Provider\ImageProviderInterface" instead.</deprecated>
</service>
<service id="Sonata\MediaBundle\Provider\ImageProviderInterface" alias="sonata.media.provider.image"/>
```
but if you think they don't belong to this PR, we could try to manage them in a new one.

jerome-fix commented:
> This will be release as a minor version because it's a new feature. 3.27.

OK. I understand better! Thank you.

jerome-fix commented:
No news ?

Jérôme.

phansys commented:
> No news ?
> 
> Jérôme.

I left a request some days ago (https://github.com/sonata-project/SonataMediaBundle/pull/1768#pullrequestreview-468077607).
Please, check the [contribution guide](https://github.com/sonata-project/SonataMediaBundle/blob/3.x/CONTRIBUTING.md#deprecation) in order to update the `UPGRADE-3.x.md` file.
If you need support, don't hesitate to ask.

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

jerome-fix commented:
> I left a request some days ago ([#1768 (review)](https://github.com/sonata-project/SonataMediaBundle/pull/1768#pullrequestreview-468077607)).
> Please, check the [contribution guide](https://github.com/sonata-project/SonataMediaBundle/blob/3.x/CONTRIBUTING.md#deprecation) in order to update the `UPGRADE-3.x.md` file.
> If you need support, don't hesitate to ask.

Sorry for the delay.

I have updated the description of PR with the "Added" and "Deprecated" items.
Is this what it was about?

I also rebased the PR on the 3.x branch.

Jérôme.

phansys commented:
>I have updated the description of PR with the "Added" and "Deprecated" items.
Is this what it was about?

The changelog section in the PR description is fine.
Since your changes include deprecations, I'm asking for the required entry at [this document](
https://github.com/sonata-project/SonataMediaBundle/blob/1391476de86cf75445b9ca43972271be033f3d7d/UPGRADE-3.x.md#upgrade-from-3x-to-3x).

VincentLanglet commented:
@jerome-fix Are you still interested in this PR ?

jordisala1991 commented:
Already done on master.
